### PR TITLE
Feature/weather fields

### PR DESCRIPTION
**Pull request by sebastinto**
## Summary
This PR enhances the `weather` command with customizable field selection, allowing users to specify exactly which weather information they want to see.

### Key Features
 - **Customizable Field Selection**: Users can now tailor the weather output by specifying desired fields (e.g. `weather denver -temp -humidity`).
 - **Extensive & Extensible Data Points**: 38 weather fields organized into categories (current, forecast, air quality, astronomy) were implemented. Full integration should be easy to implement in the future if desired.
 - **Weather Command Helper**: The `weather list` command displays all available fields and their categories.
 - **Robust Parsing**: Improved command parsing with detailed error messages for invalid syntax or field combinations.
 - **Internationalization**: Full support for 6 languages across all new functionalities.
 - **Thorough Testing**: Comprehensive coverage with 6 test suites addressing various scenarios and edge cases + manual testing performed.
 
 **Examples**: `weather london -temp -humidity`, `weather paris -uv -wind`, `weather winston-salem -sunrise -sunset`,  `weather list`

### Technical Changes
While the primary goal was to maintain consistency with existing codebase patterns, the complexity of the customizable field selection feature necessitated strategic deviation in specific areas:
- **Ktor Adoption**: Introduced for its native coroutine support, enabling seamless asynchronous operations for API calls. Ktor's built-in content negotiation simplifies JSON handling, and its structured exception handling integrates effectively with the new field validation system.
- **kotlinx.serialization Integration**: Chosen for compile-time safety in data models and automatic null handling for optional API fields. This significantly reduces boilerplate code, especially when managing the 38 field structures required for selective display.
- **More Granular Code Organization**: To manage the complexity introduced by field selection, specialized components have been organized into distinct files (`WeatherValidation.kt`, `WeatherField.kt`, `WeatherModels.kt`, `WeatherRenderer.kt`, etc.), keeping the main `Command.kt` file focused on its core responsibilities.

These changes were applied judiciously, only where feature requirements provided a clear benefit, while diligently preserving the codebase's established conventions elsewhere.

### Testing
In addition to the unit tests, please review the screenshots below:

yantra_weather_default
yantra_weather_args
yantra_weather_hyphenated_location
yantra_weather_list
yantra_weather_errors
yantra_api_error_config
yantra_api_error_location

### CODE REVIEW

**Comment by sebastinto** [comment]
Any comments, suggestions, questions are welcome. As mentioned in the PR, I tried to follow the existing coding style as much as possible but in a few instances I felt like the features required a different approach. Happy to discuss!

I tried to cover as much ground as possible during testing but obviously another set of eyes would be appreciated to catch edge cases I may have overlooked. Thanks!

File: app/src/main/java/com/coderGtm/yantra/commands/weather/WeatherModels.kt [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ * Complete data models for WeatherAPI.com response.
+ *
+ * Note that these models map the full API response, but only a subset of fields are currently exposed
+ * to users. See [WeatherField].kt for more info and instructions on adding new user-accessible
+ * fields.
+ *
+ * [https://github.com/Kotlin/kotlinx.serialization/issues/2844](More info) about why the
+ * opt-in annotation is needed.
+ */
+@file:OptIn(kotlinx.serialization.InternalSerializationApi::class)
```
[the remark this answers is not in the document]
Comment (sebastinto) [reply app/src/main/java/com/coderGtm/yantra/commands/weather/WeatherModels.kt:11]: Copilot should read the code comments 😜

File: app/src/main/java/com/coderGtm/yantra/commands/weather/Command.kt [hunk 2]
```diff
@@ -18,26 +22,40 @@ class Command(terminal: Terminal) : BaseCommand(terminal) {
...
+            install(Logging) {
+                logger = object : io.ktor.client.plugins.logging.Logger {
+                    override fun log(message: String) {
+                        Log.d("HTTP call", message)
+                    }
+                }
+                level = LogLevel.ALL
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (sebastinto) [reply app/src/main/java/com/coderGtm/yantra/commands/weather/Command.kt:45]: > Consider implementing a singleton pattern or dependency injection to reuse the HttpClient across the application

Yeah I know... 😏 

`// todo: ideally this would be a singleton that is re-used across the app`

I did not want to be more disruptive to current architecture than I am more than happy to hoist it up for reuse. LMK if you're cool with the code as-is or want to follow Copilot's suggestion @coderGtm!

File: app/src/main/java/com/coderGtm/yantra/commands/weather/Helper.kt [hunk 3]
```diff
@@ -1,82 +1,170 @@
...
+
+    val apiKey = command.terminal.activity.packageManager.getApplicationInfo(
+        command.terminal.activity.packageName,
+        PackageManager.GET_META_DATA
+    ).metaData.getString("WEATHER_API_KEY")
+
+    val url =
+        "https://api.weatherapi.com/v1/forecast.json?key=$apiKey&q=$location&lang=$langCode&aqi=yes"
+
+    CoroutineScope(Dispatchers.Main).launch {
```
[the remark this answers is not in the document]
Comment (sebastinto) [reply app/src/main/java/com/coderGtm/yantra/commands/weather/Helper.kt:44]: Fair enough. Not sure using `viewModelScope` or `lifecycleScope` really makes sense considering the app architecture but we can certainly do something like that:

``` kotlin
// class level
private var weatherJob: Job? = null

// function level
weatherJob?.cancel()
    weatherJob = CoroutineScope(Dispatchers.Main).launch {
        try {
            ensureActive()
            val weather = withContext(Dispatchers.IO) {
                httpClient.get(url).body<WeatherResponse>()
            }
            handleResponse(weather, args, command)
        } catch (e: Exception) {
            if (e is CancellationException) return@launch
            handleKtorError(e, command)
        }
    }
```

Thoughts?

File: app/src/main/java/com/coderGtm/yantra/commands/weather/Command.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (sebastinto) [reply app/src/main/java/com/coderGtm/yantra/commands/weather/Command.kt:45]: Which would look something like this:
```kotlin
/**
 * Provides a singleton HttpClient instance for the application.
 * This ensures efficient resource usage and connection pool reuse across all network requests.
 */
object HttpClientProvider {

    val client: HttpClient by lazy {
        HttpClient(Android) {
            expectSuccess = true

            install(ContentNegotiation) {
                json(Json {
                    ignoreUnknownKeys = true
                    coerceInputValues = true
                })
            }

            if (BuildConfig.DEBUG) {
                install(Logging) {
                    logger = object : io.ktor.client.plugins.logging.Logger {
                        override fun log(message: String) {
                            Log.d("HTTP call", message)
                        }
                    }
                    level = LogLevel.ALL
                }
            }
        }
    }
}

// example usage
HttpClientProvider.client.get(url).body<WeatherResponse>()
```

**Review by coderGtm — APPROVED** [review]

**Comment by coderGtm** [comment]
@sebastinto I still encounter the ktor exceptions like these:
`io.ktor.serialization.JsonConvertException: Illegal input: Fields [icon, code] are required for type with serial name 'com.coderGtm.yantra.commands.weather.Condition', but they were missing at path: $.current.condition`
